### ISSUE TITLE
unwrap the result of partial_cmp for cmp

### ISSUE THREAD

**Issue by shionryuu** [rust-lang/rust-clippy#11353]
### Summary

unwrap the result of partial_cmp for cmp may also be a choice

```
error: incorrect implementation of `partial_cmp` on an `Ord` type
  --> src/f64_heap.rs:90:1
   |
90 | /  impl PartialOrd for NonNanF64 {
91 | |      // #[allow(clippy::incorrect_partial_ord_impl_on_ord_type)]
92 | |      fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
   | | _____________________________________________________________-
93 | ||         self.0.partial_cmp(&other.0)
94 | ||     }
   | ||_____- help: change this to: `{ Some(self.cmp(other)) }`
95 | |  }
   | |__^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#incorrect_partial_ord_impl_on_ord_type
   = note: `-D clippy::incorrect-partial-ord-impl-on-ord-type` implied by `-D warnings`
```

### Reproducer

I tried this code:

```rust
use std::cmp::Ordering;

#[derive(PartialEq)]
struct NonNanF64(f64);

impl Eq for NonNanF64 {}

impl PartialOrd for NonNanF64 {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.0.partial_cmp(&other.0)
    }
}

impl Ord for NonNanF64 {
    fn cmp(&self, other: &Self) -> Ordering {
        self.partial_cmp(other).unwrap()
    }
}
```

I expected to see this happen: pass clippy

Instead, this happened: clippy fail

### Version

```text
rustc 1.73.0-nightly (d06ca0ffa 2023-08-18)
binary: rustc
commit-hash: d06ca0ffaf4ac72732665f99dd9ad962194cd0b3
commit-date: 2023-08-18
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 17.0.0
```

**Comment by Centri3** [comment]
Unwrapping in `Ord` is a pretty bad idea. When a type implements both `Ord` and `PartialOrd`, it's generally assumed that `cmp` is infallible and `partial_cmp` just wraps it in `Some`, for the case where only `PartialOrd` is implemented it's a fallible comparison. While that's not the actual purpose for `Ord` that's generally what you see in most crates. Clippy shouldn't allow this regardless, but perhaps with a clearer message

**Comment by Centri3** [comment]
Yes, imo that should be rewritten to not have `Ord` in the first place. That's an orthogonal problem to this lint though so an issue for that would warrant a separate issue + lint
